### ISSUE TITLE
collectstatic/whitenoise failure

### ISSUE THREAD

**Issue by dbinetti** [heroku/heroku-buildpack-python#653]
I noticed that `collectstatic` stopped working automatically.  

```
Total 3 (delta 2), reused 0 (delta 0)
remote: Compressing source files... done.
remote: Building source:
remote: 
remote: -----> Python app detected
remote:        Skipping installation, as Pipfile.lock hasn't changed since last deploy.
remote: -----> $ python manage.py collectstatic --noinput
remote:        Traceback (most recent call last):
remote:          File "manage.py", line 11, in <module>
remote:            execute_from_command_line(sys.argv)
remote:          File "/app/.heroku/python/lib/python3.6/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
remote:            utility.execute()
remote:          File "/app/.heroku/python/lib/python3.6/site-packages/django/core/management/__init__.py", line 317, in execute
remote:            settings.INSTALLED_APPS
remote:          File "/app/.heroku/python/lib/python3.6/site-packages/django/conf/__init__.py", line 56, in __getattr__
remote:            self._setup(name)
remote:          File "/app/.heroku/python/lib/python3.6/site-packages/django/conf/__init__.py", line 43, in _setup
remote:            self._wrapped = Settings(settings_module)
remote:          File "/app/.heroku/python/lib/python3.6/site-packages/django/conf/__init__.py", line 106, in __init__
remote:            mod = importlib.import_module(self.SETTINGS_MODULE)
remote:          File "/app/.heroku/python/lib/python3.6/importlib/__init__.py", line 126, in import_module
remote:            return _bootstrap._gcd_import(name[level:], package, level)
remote:          File "<frozen importlib._bootstrap>", line 994, in _gcd_import
remote:          File "<frozen importlib._bootstrap>", line 971, in _find_and_load
remote:          File "<frozen importlib._bootstrap>", line 941, in _find_and_load_unlocked
remote:          File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
remote:          File "<frozen importlib._bootstrap>", line 994, in _gcd_import
remote:          File "<frozen importlib._bootstrap>", line 971, in _find_and_load
remote:          File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
remote:        ModuleNotFoundError: No module named 'settings'
remote: 
remote:  !     Error while running '$ python manage.py collectstatic --noinput'
remote:        See traceback above for details.
remote: 
remote:        You may need to update application code to resolve this error.
remote:        Or, you can disable collectstatic for this application:
remote: 
remote:           $ heroku config:set DISABLE_COLLECTSTATIC=1
remote: 
remote:        https://devcenter.heroku.com/articles/django-assets
remote:  !     Push rejected, failed to compile Python app.
remote: 
remote:  !     Push failed
remote: Verifying deploy...
remote: 
remote: !	Push rejected to barberscore-api-staging.
remote: 
To https://git.heroku.com/barberscore-api-staging.git
 ! [remote rejected]   HEAD -> master (pre-receive hook declined)
error: failed to push some refs to 'https://git.heroku.com/barberscore-api-staging.git'
```

If I set DISABLE_COLLECTSTATIC=1 then it will build.   I should mention that I've set my config so that I can access commands with `django-admin` instead of `python manage.py `; not sure if that's relevant.

Whitenoise has also failed, not sure if related.  And not sure how to debug...

**Comment by ryanhiebert** [comment]
The buildpack uses the presence of `Django` in the requirements and `manage.py` at the root of the repository to determine whether to run `collectstatic`, and it uses `python manage.py collectstatic` to do it, so I'm not sure there's a way around having the manage.py file existing right now.

I also wanted to just use `django-admin` instead of `python manage.py`. And I also wanted to have the command be called `django` instead of `django-admin`, for brevity and obviousness (my own perception of it). That's why I created the [`django-cmd`](https://pypi.python.org/pypi/django-cmd) package. You'll still need the manage.py file on Heroku, as I mentioned earlier, but I like running `django migrate` as my migration command.

**Comment by kennethreitz** [comment]
What happens when you run `$ heroku run python manage.py collectstatic --noinput`?

**Comment by dbinetti** [comment]
@ryanhiebert Thanks, but I don't need to remove `manage.py`; I just mentioned it in case it was relevant.

@kennethreitz Running that works as expected.  Also, why was this closed?  Is it not relevant to the buildpack?  The issue is still happening for me.

I should also mention that I have my settings in a settings package (ie, settings directory with `__init__.py`) and run my DJANGO_MODULE_SETTINGS to the path to the settings I need: such as `settings.dev`, `settings.staging`, etc.  Again, in case it's relevant.

**Comment by kennethreitz** [comment]
@dbinetti i think this is the cause of your problem. It sounds like you may need to either customize the buildpack to suit your needs (not recommended) or use a single settings file (recommended).

**Comment by dbinetti** [comment]
Not entirely on topic, but what is best practice on that for 12-factor?  I already use environment variables for config, but there are still significant differences (particularly on test settings.). Just a lot of branching conditionals in the settings file or is there another common solution?

**Comment by kennethreitz** [comment]
Ideally, you have a single settings file that would branch based on environment variables present.

**Comment by dbinetti** [comment]
ty

**Comment by kennethreitz** [comment]
The goal is to branch as little as possible, of course :)

**Comment by dbinetti** [comment]
Aaaaaand now that I spent a day refactoring my settings...  same issue!  :-(

If this were more widespread I suppose we'd have heard it by now.  But still hoping something can be done about it.

```
-----> Python app detected
       Skipping installation, as Pipfile.lock hasn't changed since last deploy.
-----> $ python manage.py collectstatic --noinput
       Traceback (most recent call last):
         File "manage.py", line 11, in <module>
           execute_from_command_line(sys.argv)
         File "/app/.heroku/python/lib/python3.6/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
           utility.execute()
         File "/app/.heroku/python/lib/python3.6/site-packages/django/core/management/__init__.py", line 317, in execute
           settings.INSTALLED_APPS
         File "/app/.heroku/python/lib/python3.6/site-packages/django/conf/__init__.py", line 56, in __getattr__
           self._setup(name)
         File "/app/.heroku/python/lib/python3.6/site-packages/django/conf/__init__.py", line 43, in _setup
           self._wrapped = Settings(settings_module)
         File "/app/.heroku/python/lib/python3.6/site-packages/django/conf/__init__.py", line 106, in __init__
           mod = importlib.import_module(self.SETTINGS_MODULE)
         File "/app/.heroku/python/lib/python3.6/importlib/__init__.py", line 126, in import_module
           return _bootstrap._gcd_import(name[level:], package, level)
         File "<frozen importlib._bootstrap>", line 994, in _gcd_import
         File "<frozen importlib._bootstrap>", line 971, in _find_and_load
         File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
       ModuleNotFoundError: No module named 'sets'
 !     Error while running '$ python manage.py collectstatic --noinput'.
       See traceback above for details.
       You may need to update application code to resolve this error.
       Or, you can disable collectstatic for this application:
          $ heroku config:set DISABLE_COLLECTSTATIC=1
       https://devcenter.heroku.com/articles/django-assets
 !     Push rejected, failed to compile Python app.
 !     Push failed
```

```
(barberscore-api)bash-3.2$ staging run python manage.py collectstatic --noinput
Running python manage.py collectstatic --noinput on ⬢ barberscore-api-staging... up, run.2295 (Hobby)
Copying '/app/.heroku/python/lib/python3.6/site-packages/django/contrib/admin/static/admin/img/icon-unknown-alt.svg'
...
171 static files copied to '/app/project/staticfiles'.
(barberscore-api)bash-3.2$ 
```

**Comment by kennethreitz** [comment]
Do you have PYTHONHOME set on your app?

**Comment by cclauss** [comment]
__ModuleNotFoundError: No module named ‘sets__.  https://docs.python.org/2/library/sets.html has been replaced with the builtin https://docs.python.org/3.6/library/stdtypes.html#frozenset

**Comment by kennethreitz** [comment]
Ah, a recent builtpack change is overwriting that. I’ll have to change that to only set it if it’s under by the user.

Sorry for the headache!

**Comment by dbinetti** [comment]
No worries.  Figured it was something like that.

**Comment by kennethreitz** [comment]
I’ll try to have a fix out on Monday!

**Comment by dbinetti** [comment]
Hey @kennethreitz is this still on track for release today?  I need to push some changes and am blocked on this.

Also, if this is expected to be a problem in the future I can adjust.  But unless some underlying architectural changes have been made it really is a lot simpler to set the PYTHONPATH to my project home dir.  Thanks

**Comment by dbinetti** [comment]
Actually @kennethreitz  I just realized I can use a tagged build pack from GitHub, so I'll just do that until this gets resolved.  Meaning: no rush.

**Comment by kennethreitz** [comment]
fixed by #655

**Comment by kennethreitz** [comment]
released!
